common/lib/xmodule/xmodule/docs/decisions/0001-move-due-dates-to-db.rst

```diff
@@ -0,0 +1,52 @@
+1. Move Due Dates to Relational Database
+----------------------------------------
+
+Status
+------
+
+Proposed
+
+Context
+-------
+
+We want an authoritative and readily-accessible place to track due dates for sections/units in a course.  We also want to override due dates per learner, and record the reason for the override. The current implementation stores due dates in the XBlock (in mongodb). If you want to know all of the due dates in a course, you have to walk the entire course structure and pull out each date, which is an inefficient operation.
+
+There is currently a way to override due dates per learner, using `IDDE <https://github.com/mitodl/ccx-idde-overrides-slides/blob/master/markdown/slides.md#individual-due-date-extensions-idde>`_, but this relies on a generic field-override system which doesn't know anything about the underlying data, and enabling this feature results in a database query for every xblock access, for every course on the platform. IDDE also does not allow for an audit trail on due date extensions, nor does it allow instructors to record a reason for the extension. 
+
+
+Decisions
+---------
+
+1. Create a new pluggable Django app responsible for dates and date overrides
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This approach anticipates other date-related functions in the future, such as APIs for returning a calendar view of upcoming dates.
+
+- The app will contain at least these models:
+    + ``ContentDate``: mapping content ids to dates
+    + ``UserContentDate``: mapping user ids to ``ContentDates``. It will also record an audit trail for the override.
+- The app will contain a custom XBlock field that will be a drop-in replacement for the existing Date XBlock field.
+    + If the date exists in the relational database (whether in UserContentDate or ContentDate), it'll use that date; otherwise, it'll use the one stored in the XBlock.
```

Comment:
I'd be concerned that making this opt-in would introduce incompatibilities if an XBlock out there declared its `due` field the [old-fashioned way](https://github.com/edx/edx-ora2/blob/master/openassessment/xblock/lms_mixin.py#L35). We can do it at the field data layer though, without having the XBlock know about it.

I mean, honestly, I really dislike the fact that we use a field for this at all instead of having a service that can say "What's my due status?" or some such. I'd like to add a service that explicitly lets you ask all kinds of scheduling things eventually. But I think a read-only shim at the field data service layer will give us the broadest compatibility while maintaining our ability to make performant queries.

Reply:
+1 Agree that for types of data that are not specific to an XBlock, a field does impose limitations e.g. needing to put helper code in the XBlock class/mixins and everything is in the same namespace which is messy.

A service will also allow XBlocks/other features/plugins to specify customs events and due dates can be one of them.